### PR TITLE
copy k8s.io/kubernetes/pkg/util/netsets to k8s.io/util/netsets

### PR DESCRIPTION
**Pull request by andrewsykim**
Once merged I will update k8s.io/utils vendor and replace k8s.io/kubernetes/pkg/util/net. 

/assign @lavalamp

### CODE REVIEW

**Comment by thockin** [comment]
Can you talk about why to promote this?    Moving to top-level utils means a more thorough API examination.

File: net/net.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// IsIPv6CIDR returns if cidr is IPv6.
+// This assumes cidr is a valid CIDR.
+func IsIPv6CIDR(cidr string) bool {
+	ip, _, _ := net.ParseCIDR(cidr)
+	return IsIPv6(ip)
+}
+
+// FilterIncorrectIPVersion filters out the incorrect IP version case from a slice of IP strings.
```
[the remark this answers is not in the document]
Comment (lavalamp) [reply net/net.go]: The API is bizarre, agreed. If you want to move it here, I suggest changing to something like:

```
func SplitIPsByFamily(ipList []string) (ipv4List []string, ipv6List []string) {
...
func SplitCIDRsByFamily(cidrList []string) (ipv4List []string, ipv6List []string) {
```

That's still strange but at least the interface isn't WTF worthy.

Overall I agree with Tim, if it's only used one place, just move it there.

[the remark this answers is not in the document]
Comment (andrewsykim) [reply net/net.go]: A few things:
* this is just a copy/paste from `kubernetes/kubernetes`, I don't think we should be changing the API in the same PR as moving it to a new repo, happy to follow up in a new PR and make those changes though
* netsets is actually imported from many places (GCE provider being one of them):
vider/providers
```
$ git grep netsets
pkg/api/service/util.go:        netsets "k8s.io/kubernetes/pkg/util/net/sets"
pkg/api/service/util.go:// IsAllowAll checks whether the netsets.IPNet allows traffic from 0.0.0.0/0
pkg/api/service/util.go:func IsAllowAll(ipnets netsets.IPNet) bool {
pkg/api/service/util.go:func GetLoadBalancerSourceRanges(service *api.Service) (netsets.IPNet, error) {
pkg/api/service/util.go:        var ipnets netsets.IPNet
pkg/api/service/util.go:                ipnets, err = netsets.ParseIPNets(specs...)
pkg/api/service/util.go:                ipnets, err = netsets.ParseIPNets(specs...)
pkg/api/service/util_test.go:   netsets "k8s.io/kubernetes/pkg/util/net/sets"
pkg/api/service/util_test.go:   checkOK := func(v string) netsets.IPNet {
pkg/api/service/util_test.go:           ipnets, err := netsets.ParseIPNets(cidrs...)
pkg/api/v1/service/util.go:     netsets "k8s.io/kubernetes/pkg/util/net/sets"
pkg/api/v1/service/util.go:// IsAllowAll checks whether the netsets.IPNet allows traffic from 0.0.0.0/0
pkg/api/v1/service/util.go:func IsAllowAll(ipnets netsets.IPNet) bool {
pkg/api/v1/service/util.go:func GetLoadBalancerSourceRanges(service *v1.Service) (netsets.IPNet, error) {
pkg/api/v1/service/util.go:     var ipnets netsets.IPNet
pkg/api/v1/service/util.go:             ipnets, err = netsets.ParseIPNets(specs...)
pkg/api/v1/service/util.go:             ipnets, err = netsets.ParseIPNets(specs...)
pkg/api/v1/service/util_test.go:        netsets "k8s.io/kubernetes/pkg/util/net/sets"
pkg/api/v1/service/util_test.go:        checkOK := func(v string) netsets.IPNet {
pkg/api/v1/service/util_test.go:                ipnets, err := netsets.ParseIPNets(cidrs...)
pkg/cloudprovider/providers/gce/gce_loadbalancer.go:    netsets "k8s.io/kubernetes/pkg/util/net/sets"
pkg/cloudprovider/providers/gce/gce_loadbalancer.go:    ipn   netsets.IPNet
pkg/cloudprovider/providers/gce/gce_loadbalancer.go:    lbSrcRngsFlag.ipn, err = netsets.ParseIPNets([]string{"130.211.0.0/22", "35.191.0.0/16", "209.85.152.0/22", "209.85.204.0/22"}...)
pkg/cloudprovider/providers/gce/gce_loadbalancer.go:            c.ipn = make(netsets.IPNet)
pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:   netsets "k8s.io/kubernetes/pkg/util/net/sets"
pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:func (gce *GCECloud) firewallNeedsUpdate(name, serviceName, region, ipAddress string, ports []v1.ServicePort, sourceRanges netsets.IPNet) (exists bool, needsUpdate bool, err error) {
pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:   actualSourceRanges, err := netsets.ParseIPNets(fw.SourceRanges...)
pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:func (gce *GCECloud) createFirewall(svc *v1.Service, name, region, desc string, sourceRanges netsets.IPNet, ports []v1.ServicePort, hosts []*gceInstance) error {
pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:func (gce *GCECloud) updateFirewall(svc *v1.Service, name, region, desc string, sourceRanges netsets.IPNet, ports []v1.ServicePort, hosts []*gceInstance) error {
pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:func (gce *GCECloud) firewallObject(name, region, desc string, sourceRanges netsets.IPNet, ports []v1.ServicePort, hosts []*gceInstance) (*compute.Firewall, error) {
pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go:      netsets "k8s.io/kubernetes/pkg/util/net/sets"
pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go:      ipnet, err := netsets.ParseIPNets("0.0.0.0/0")
pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go:      wrongIpnet, err := netsets.ParseIPNets("1.0.0.0/10")
pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go:              ipnet        netsets.IPNet
pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go:      ipnet, err := netsets.ParseIPNets("10.0.0.0/20")
pkg/util/net/sets/doc.go:// It is expected that callers will alias this import as "netsets" i.e. import
pkg/util/net/sets/doc.go:// netsets "k8s.io/kubernetes/pkg/util/net/sets"
```
* we could move it to `pkg/proxy` but that would defeat the purpose as the GCE provider will still end up depending on `k8s.io/kubernetes`.

[the remark this answers is not in the document]
Comment (andrewsykim) [reply net/net.go]: ahh, realizing now that you were referring to `net` package only and not `net/sets`, are we okay with moving `net/sets` here though?

**Comment by lavalamp** [comment]
I was referring to only this function and the one below? Maybe Tim meant
the whole package :)

**Comment by andrewsykim** [comment]
ahh okay, sorry for misunderstanding. I updated the PR to only move `k8s.io/kubernetes/pkg/util/net/sets` -> `k8s.io/utils/netsets` as netsets is used in a lot of places. As for changes to `k8s.io/kubernetes/pkg/util/net/sets` I will try to address some concerns mentioned in `kubernetes/kubernetes` in a follow up PR :)

File: netsets/doc.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+// This package contains hand-coded set implementations that should be similar
+// to the autogenerated ones in pkg/util/sets.
+// We can't simply use net.IPNet as a map-key in Go (because it contains a
+// []byte).
+// We could use the same workaround we use here (a string representation as the
+// key) to autogenerate sets.  If we do that, or decide on an alternate
+// approach, we should replace the implementations in this package with the
+// autogenerated versions.
+// It is expected that callers will alias this import as "netsets" i.e. import
+// netsets "k8s.io/utils/netsets"
```
Comment (stewart-yu) [inline netsets/doc.go]: How about fix those comments? outdate now

Comment (andrewsykim) [reply netsets/doc.go]: done, thanks

**Comment by andrewsykim** [comment]
@lavalamp @thockin @stewart-yu I've made some changes based on feedback so we're only migrating `pkg/util/net/sets` here. Once merged I will update vendor in `k8s.io/kubernetes`, replace `k8s.io/kubernetes/pkg/util/net/sets` with `k8s.io/utils/netsets`. Any objections to move forward?

**Comment by lavalamp** [comment]
Maybe we can document with the package that the behavior is undefined if the CIDRs overlap? I don't know what this is used for but it is confusing to me that it doesn't verify that, given what it is called.

I am fine with moving it here, but maybe Tim has a higher bar.

File: netsets/README.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Netsets
+
+Netsets is a util package providing a map of "net.IPNet". It is used in various 
+placed in k8s.io/kubernetes to manage IP addresses for services and loadbalancers. 
+
+This package was originally copied over from `k8s.io/kubernetes/pkg/util/net/sets`.
```
Comment (stewart-yu) [inline netsets/README.md]: seems should change the `lib` parts in `https://github.com/kubernetes/utils/blob/master/README.md`, not add one :grin:

Comment (andrewsykim) [reply netsets/README.md]: ah, didn't see that there, good catch!

Comment (andrewsykim) [reply netsets/README.md]: fixed!

**Comment by andrewsykim** [comment]
@thockin any objections to move forward with this? I've changed the PR so we're only migrating `k8s.io/kubernetes/pkg/util/net/sets` and I understand that the API for `netsets` is a bit odd but I'm hestitant to try to change the API here as it'll introduce more changes when we vendor it back into `k8s.io/kubernetes`.

**Comment by andrewsykim** [comment]
Closing this as @mcrute is taking it (with commit history preserved)

**Pull request closed without merging**